docker/go-tf-prepare/pkg/azure/azure.go

```diff
@@ -142,6 +147,64 @@ func CreateStorageAccount(ctx context.Context, cred azcore.TokenCredential, conf
 	return err
 }
 
+func registerResourceProviderIfNeeded(ctx context.Context, cred azcore.TokenCredential, config azureConfig, resourceProviderNamespace string) error {
+	subscriptionID := config.SubscriptionID
+	log, err := logr.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	client, err := armresources.NewProvidersClient(subscriptionID, cred, &arm.ClientOptions{})
+	if err != nil {
+		log.Error(err, "armresources.NewProvidersClient")
+		return err
+	}
+
+	res, err := client.Get(ctx, resourceProviderNamespace, &armresources.ProvidersClientGetOptions{})
+	if err != nil {
+		log.Error(err, "client.Get")
+		return err
+	}
+
+	if *res.RegistrationState == "Registered" {
+		log.Info("Azure Resource Provider already registered", "resourceProviderNamespace", resourceProviderNamespace, "registrationState", *res.RegistrationState)
+		return nil
+	}
+
+	regRes, err := client.Register(ctx, resourceProviderNamespace, &armresources.ProvidersClientRegisterOptions{})
+	if err != nil {
+		log.Error(err, "client.Register")
+		return err
+	}
+
+	log.Info("Registering Azure Resource Provider", "resourceProviderNamespace", resourceProviderNamespace, "registrationState", *regRes.RegistrationState)
+
+	if *regRes.RegistrationState == "Registered" {
+		log.Info("Azure Resource Provider registered", "resourceProviderNamespace", resourceProviderNamespace, "registrationState", *regRes.RegistrationState)
+		return nil
+	}
+
+	currentRegistrationState := ""
+	for i := 1; i < 10; i++ {
+		res, err := client.Get(ctx, resourceProviderNamespace, &armresources.ProvidersClientGetOptions{})
+		if err != nil {
+			log.Error(err, "client.Get")
+			return err
+		}
+		currentRegistrationState = *res.RegistrationState
+		if currentRegistrationState == "Registered" {
+			log.Info("Azure Resource Provider registered", "resourceProviderNamespace", resourceProviderNamespace, "registrationState", currentRegistrationState)
+			return nil
+		}
+		time.Sleep(time.Duration(i*5) * time.Second)
```

Comment:
You should probably bail if status != "Registering" or whatever it normally is while it is pondering. The registration can be minutes long, so you might want to print some feedback.

Reply:
@bittrance fixed!